src/main/java/com/cburch/logisim/fpga/hdlgenerator/HDLPorts.java

```diff
@@ -0,0 +1,192 @@
+/*
+ * Logisim-evolution - digital logic design tool and simulator
+ * Copyright by the Logisim-evolution developers
+ *
+ * https://github.com/logisim-evolution/
+ *
+ * This is free software released under GNU GPLv3 license
+ */
+
+package com.cburch.logisim.fpga.hdlgenerator;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import com.cburch.logisim.data.Attribute;
+import com.cburch.logisim.data.AttributeSet;
+import com.cburch.logisim.data.BitWidth;
+import com.cburch.logisim.instance.Port;
+
+
+public class HDLPorts {
+ 
+  public static final String CLOCK = "clock";
+  public static final String TICK = "tick";
+  public static final String PULL_DOWN = "fixed_pull_down";
+  public static final String PULL_UP = "fixed_pull_up";
+
+  private class PortInfo {
+
+    private final String myPortType;
+    private final String myName;
+    private final int myNrOfBits;
+    private final int myComponentPinId;
+    private final String myFixedMap;
+    private boolean mySinglePinException = false;
+    private Attribute<?> myBitWidthAttribute = null;
+    private boolean myPullToZero = true;
+    private boolean isClock = false;
+
+    public PortInfo(String type, String name, int nrOfBits, String fixedMap) {
+      myPortType = type;
+      myName = name;
+      myNrOfBits = nrOfBits;
+      myComponentPinId = -1; // fixed map
+      myFixedMap = fixedMap;
+    }
+
+    public PortInfo(String type, String name, int nrOfBits, int compPinId) {
+      this(type, name, nrOfBits, compPinId, null);
+      mySinglePinException = false;
+    }
+    
+    public PortInfo(String type, String name, int nrOfBits, int compPinId, boolean pullToZero) {
+      this(type, name, nrOfBits, compPinId, null);
+      mySinglePinException = false;
+      myPullToZero = pullToZero;
+    }
+    
+    public PortInfo(String type, String name, int nrOfBits, int compPinId, Attribute<?> nrOfBitsAttr) {
+      myPortType = type;
+      myName = name;
+      myNrOfBits = nrOfBits;
+      myComponentPinId = compPinId;
+      myFixedMap = null;
+      mySinglePinException = true;
+      myBitWidthAttribute = nrOfBitsAttr;
+    }
+    
+    int getNrOfBits(AttributeSet attrs) {
+      if (mySinglePinException) {
+        if (!attrs.containsAttribute(myBitWidthAttribute)) throw new IllegalArgumentException("Bitwidth attribute not found");
+        final var value = attrs.getValue(myBitWidthAttribute);
+        var nrOfBits = 0;
+        if (value instanceof BitWidth) {
+          nrOfBits = ((BitWidth) value).getWidth();
+        } else if (value instanceof Integer) {
+          nrOfBits = (int) value;
+        } else throw new IllegalArgumentException("No Bitwidth or Integer value");
```

Comment:
I think exception message is bit cryptic (esp. `No Bitwidth`). Mind rephrasing a bit?

Reply:
done